prov/verbs/src/verbs_mr.c

```diff
@@ -76,9 +80,36 @@ struct ibv_mr *vrb_mr_ibv_reg_dmabuf_mr(struct ibv_pd *pd, const void *buf,
 		return NULL;
 
 	offset = (uintptr_t)buf - (uintptr_t)base;
-	return ibv_reg_dmabuf_mr(pd, offset, len, (uint64_t)buf/* iova */,
-				 (int)(uintptr_t)handle/* dmabuf fd */,
-				 vrb_access);
+	mr = ibv_reg_dmabuf_mr(pd, offset, len, (uint64_t)buf/* iova */,
+			       (int)(uintptr_t)handle/* dmabuf fd */,
+			       vrb_access);
+	if (!mr && failover_needed)
+		goto failover;
+
+	/* no more failover once dmabuf mr is proven working */
+	failover_needed = 0;
+	return mr;
+
+failover:
+	if (vrb_gl_data.peer_mem_support) {
+		saved_errno = errno;
+		mr = ibv_reg_mr(pd, (void *)buf, len, vrb_access);
+		if (!mr) {
+			errno = saved_errno;
+			failover_cnt = 0;
+			return NULL;
+		}
+
+		/* disable dmabuf mr after consecutive successful failovers */
+		failover_cnt++;
+		FI_INFO(&vrb_prov, FI_LOG_MR, "Dmabuf mr failover succeeded.\n");
+		if (failover_cnt > 5) {
+			vrb_gl_data.dmabuf_support = 0;
+			FI_INFO(&vrb_prov, FI_LOG_MR,
+				"Failover limit reached. Disable ibv_reg_dmabuf_mr()!\n");
+		}
+	}
+	return mr;
```

Comment:
I was expecting to see a check at the top of the function to goto failover and skip calling ibv_reg_dmabuf_mr() if we ever took the failover path.

Reply:
The check is at the call site. See the second commit. Here the ibv_reg_dmabuf_mr() path is given five chances, if it always fail then the 'dmabuf_support' flag is reset and future calls will not get here at all.

Reply:
I meant if ibv_reg_dmabuf_mr() fails, but ibv_reg_mr() succeeds, we continue to try that call sequence -- letting ibv_reg_dmabuf_mr() fail before calling ibv_reg_mr().  I was thinking something more like:
```
static bool use_failover = false;

if (use_failover)
   goto failover;

mr = ibv_reg_dmabuf_mr()
if (mr)
    return mr;

use_failover = true

failover:
```
This isn't quite capturing the full state correctly, but hopefully shows the idea..  If ibv_reg_dmabuf_mr() fails the first time, do we ever expect it to work?  If so, can we mix ibv_reg_dmabuf_mr() with ibv_reg_mr()

Reply:
We still need to count for the case that the call is meant to fail -- e.g. due to incorrect parameters -- and return proper error code back to the user.  So we can't switch to permanent failover mode after ibv_reg_dmabuf_mr(0 fails. This has to wait until the failover actually succeeds. That's the logic of the current code, but with a threshold of 5 times. That can be changed to 1, though.

Reply:
I don't think I'm getting my point across clearly.  Say that ibv_reg_dmabuf_mr() fails every time, but ibv_reg_mr() works every time.  This is the actual failover that we're trying support, correct?  In this case, we always continue to call ibv_reg_dmabuf_mr() over and over even though it will never work.  Maybe this makes the point better:
```
static bool failover_enabled = true;
static size_t failover_good_cnt;
static size_t failover_bad_cnt;

if (!failover_enabled || failover_good_cnt < 5) {
    mr = ibv_reg_dmabuf_mr();
    if (mr) {
        failover_enabled = false;
        return mr;
    }
}

mr = ibv_reg_mr();
if (mr) {
    failover_good_cnt++;
    return mr;
}

if (failover_bad_cnt++ > 5)
    disable forever

```

There's likely a more elegant way of handling this.

Also, this provider has a build requirement on both ze and libibverbs.  Is that captured in the configure?  Furthermore, is this hooking provider only really usable over the verbs provider?  It looks like it.  Does it make more sense to simply integrate this code with the verbs provider directly?  What are the benefits of keeping it separate?  What happens if this is enabled, but both the verbs and shm providers are in use?

Reply:
>>Also, this provider has a build requirement on both ze and libibverbs. Is that captured in the configure?

>No yet. will fix.

A further look into this -- the provider actually doesn't have build dependency on ze and libibverbs. The ze calls are thru the wrappers defined in src/hmem_ze.c which has already handled the configuration check. No verbs calls are made from this provider.

Reply:
Here is a revised version of **vrb_mr_ibv_reg_dmabuf_mr**. Does it look better than the current one?
```
static inline
struct ibv_mr *vrb_mr_ibv_reg_dmabuf_mr(struct ibv_pd *pd, const void *buf,
                                        size_t len, int vrb_access)
{
        void *handle;
        void *base;
        uint64_t offset;
        int err;
        struct ibv_mr *mr;
        int saved_errno = 0;
        enum { TRY, ALWAYS, NEVER };
        static int failover_policy = TRY;
 
        if (failover_policy == ALWAYS)
                goto failover;
 
        err = ze_hmem_get_handle((void *)buf, &handle);
        if (err)
                return NULL;
 
        err = ze_hmem_get_base_addr((void *)buf, &base, &len);
        if (err)
                return NULL;
 
        offset = (uintptr_t)buf - (uintptr_t)base;
        mr = ibv_reg_dmabuf_mr(pd, offset, len, (uint64_t)buf/* iova */,
                               (int)(uintptr_t)handle/* dmabuf fd */,
                               vrb_access);
        if (!mr && failover_policy == TRY && vrb_gl_data.peer_mem_support) {
                saved_errno = errno;
                goto failover;
        }
 
        failover_policy = NEVER;
        return mr;

failover:
        mr = ibv_reg_mr(pd, (void *)buf, len, vrb_access);
        if (!mr) {
                if (saved_errno)
                        errno = saved_errno;
                return NULL;
        }
 
        if (failover_policy == TRY) {
                failover_policy = ALWAYS;
                FI_INFO(&vrb_prov, FI_LOG_MR,
                        "Failover on: ibv_reg_dmabuf_mr() ==> ibv_reg_mr\n");
        }
        return mr;
}
```

Reply:
Yes -- the new state enum helps make it clearer what's happening.

Reply:
Thanks! Just updated the PR with the new version.